src/transformers/pipelines/token_classification.py

```diff
@@ -212,22 +269,83 @@ def __call__(self, inputs: Union[str, List[str]], **kwargs):
                     "start": start_ind,
                     "end": end_ind,
                 }
-
-                if self.grouped_entities and self.ignore_subwords:
+                # These fields will be consumed by self.aggregate and not appear
+                # in the final result
+                if self.aggregation_strategy != AggregationStrategy.NONE:
                     entity["is_subword"] = is_subword
+                if self.aggregation_strategy == AggregationStrategy.AVERAGE:
+                    # AVERAGE needs to keep intermediate scores.
+                    entity["scores"] = score[idx]
 
                 entities += [entity]
 
-            if self.grouped_entities:
-                answers += [self.group_entities(entities)]
-            # Append ungrouped entities
-            else:
-                answers += [entities]
+            # Might be no-op for NONE strategy
+            grouped_entities = self.aggregate(entities, self.aggregation_strategy)
+            answers += [grouped_entities]
 
         if len(answers) == 1:
             return answers[0]
         return answers
 
+    def aggregate(self, entities: List[dict], aggregation_strategy: AggregationStrategy) -> List[dict]:
+        if aggregation_strategy == AggregationStrategy.NONE:
+            return entities
+        if aggregation_strategy != AggregationStrategy.SIMPLE:
+            entities = self.aggregate_words(entities, aggregation_strategy)
+        return self.group_entities(entities)
+
+    def aggregate_word(self, entities: List[dict], aggregation_strategy: AggregationStrategy) -> dict:
+        word = self.tokenizer.convert_tokens_to_string([entity["word"] for entity in entities])
+        if aggregation_strategy == AggregationStrategy.FIRST:
+            entity = entities[0]["entity"]
+            score = entities[0]["score"]
+        elif aggregation_strategy == AggregationStrategy.MAX:
+            max_entity = max(entities, key=lambda entity: entity["score"])
+            score = max_entity["score"]
+            entity = max_entity["entity"]
+        elif aggregation_strategy == AggregationStrategy.AVERAGE:
+            scores = np.stack([entity["scores"] for entity in entities])
+            average_scores = np.nanmean(scores, axis=0)
+            entity_idx = average_scores.argmax()
+            entity = self.model.config.id2label[entity_idx]
```

Comment:
umm `average_scores.argmax()` would just return the index where max score occurs. which !=entity_idx on `self.model.config.id2label`
am I missing something? shouldn't it be:
```py
entity = entities[entity_idx]["entity"] # which is already self.model.config.id2label mapped
```

Reply:
also not sure what exactly is happening in the average strategy based on the code 😵 In the code `scores` looks like a list [0.1, 0.5 , 0.1, 0.4] which makes averaging then argmaxing make no sense.

in my mind avg is:
|I-ORG | I-PER | I-PER | I-LOC|
|---|---|---|---|
|0.1      |  0.5    |   0.1   |  0.4|

first strategy -> 0.1 I-ORG
max strategy -> 0.5 I-PER
**avg strategy -> (0.3 I-PER, 0.4 I-LOC 0.1 I-ORG )->0.4 I-LOC**
I guess there can also be a sum strategy -> (0.6 I-PER, 0.4 I-LOC 0.1 I-ORG ) -> 0.6 I-PER

(btw, scores in reality are floats so we wouldn't be concerned about tie breaking (in max) etc.)

Reply:
Isn't `scores` a 2D array of dimensions (N_subwords, N_softmax)? We are averaging along subwords and then taking the  overall argmax label. This makes sense if we assume each subword carries some amount of uncorrelated signal about the entity label.

Reply:
aha just realized, it all makes sense now!
Nevermind, morning brain  🤦